blottertrax/description_provider.py

```diff
@@ -0,0 +1,71 @@
+import musicbrainzngs
+
+from blottertrax.config import Config
+from blottertrax.exceptions.description_exception import DescriptionException
+from blottertrax.helper import templates
+from blottertrax.value_objects.parsed_submission import ParsedSubmission
+
+
+class DescriptionProvider:
+    musicbrainz = None
+
+    def __init__(self):
+        config = Config()
+
+        self.musicbrainz = musicbrainzngs
+        self.musicbrainz.auth(config.MUSICBRAINZ_USER, config.MUSICBRAINZ_PASSWORD)
+        self.musicbrainz.set_useragent("BlotterTrax", "0.1", "https://github.com/martijnboers/BlotterTrax")
+
+    def get_reply(self, parsed_submission: ParsedSubmission) -> str:
+        if parsed_submission.success is False:
+            raise DescriptionException('DescriptionProvider requires the submission title')
+
+        query = '"{}" AND artist:"{}"'.format(parsed_submission.track_title, parsed_submission.artist)
+        result = self.musicbrainz.search_recordings(query=query, limit=1, )
+
+        if result['recording-count'] == 0:
+            raise DescriptionException('No recordings found for artist')
+
+        recording = result['recording-list'][0]
+        artist_ids = list(map(lambda a: a['artist']['id'], recording['artist-credit']))
+        artists = list(map(lambda a: self._get_artist_by_id(a), artist_ids))
+
+        life_span_begin = self._safe_list_get(artists[0], '?', 'life-span', 'begin')
+        life_span_end = self._safe_list_get(artists[0], 'now', 'life-span', 'end')
+        has_life_span = (life_span_end != 'now' or life_span_begin != '?')
+        has_tags = (self._safe_list_get(artists[0], False, 'tag-list') is not False)
+        has_socials = (self._safe_list_get(artists[0], False, 'url-relation-list') is not False)
+
+        if has_tags is False and has_socials is False:
+            raise DescriptionException('Neither tags nor socials found, skipping')
+
+        life_span = '' if has_life_span is False else '({} to {})'.format(life_span_begin, life_span_end)
+        tags = ', '.join(map(lambda t: t['name'], artists[0]['tag-list'][:5])) if has_tags else 'none'
+        socials = ', '.join(map(lambda u: '[{}]({})'.format(u['type'], u['target']),
+                                artists[0]['url-relation-list'])) if has_socials else 'none'
+
+        return templates.musicbrainz_artist_info.strip().format(
+            artists[0]['name'],
+            life_span,
+            recording['title'],
+            recording['release-list'][0]['date'],
+            tags,
+            socials,
+            artists[0]['id']
+        )
+
+    def _get_artist_by_id(self, artist_id: str):
+        return self.musicbrainz.get_artist_by_id(
+            id=artist_id, includes=['tags', 'ratings', 'annotation', 'url-rels', 'user-tags']
+        )['artist']
+
+    @staticmethod
```

Comment:
Should this method be defined outside this class?  I know the Soundcloud threshold check throws KeyError fairly often so might be good for that class to be able to use the same implementation without calling it on this class.

Reply:
Done